### ISSUE TITLE
Isolate Crt/Keyboard functionality

### ISSUE THREAD

**Issue by pleumann** [stefanhaustein/empire#4]
And find a way to support different terminal types (for instance via includes). The following would be good to have:

VT52 (cp/m in general)
Spectrum Next
VT100
ANSI
Line-mode (maybe, would complicate the current logic)

**Comment by stefanhaustein** [comment]
Includes won't allow parallel use (which a server needs to support), so a dynamic mode enum seems best? 
Do we really need both, VT52 and VT100?

**Comment by pleumann** [comment]
Not sure what we mean by "parallel use". Maybe we should have another video conference with the whole project committee to discuss? :)

VT52 is different from VT100. VT52 would be needed for cm/m (Spectrum Next supports extended VT52 with colors, tnylpo supports plain VT52 and is good for development and testing). VT100/ANSI would be needed for more modern cases such as a Linux or Max terminal window.

**Comment by stefanhaustein** [comment]
I think we had already discussed this in the last VC: Telnet/ssh server
should be able to serve any terminal type client.

Am Fr., 3. Juli 2020 um 17:38 Uhr schrieb pleumann <notifications@github.com

**Comment by pleumann** [comment]
Hmm... I think we may be better off keeping the old local/remote distinction.

Remote can always be VT100. Any system that can invoke telnet/ssh must be so recent that it supports VT100.

Local could use CRT when compiled with Free Pascal or Turbo Pascal 4.0+ and a VT52-based CRT clone when on the Spectrum or in another cp/m implementation (basically that means Turbo Pascal 3, most procedures are already there, which I didn't expect originally, just SetColor and some smaller stuff is missing).

I think we don't have a case where we would ever serve VT52 via network.

**Comment by stefanhaustein** [comment]
I was assuming there was some way to connect the next or cp/m to the
internet....?

How would a local game even work at all? Add a computer player? And speed
up the turns?




Am Fr., 3. Juli 2020 um 20:21 Uhr schrieb pleumann <notifications@github.com

**Comment by pleumann** [comment]
Yes, and the (well, my) idea was to make the Next the server, in which case it would serve VT100. I'm assuming any proper terminal program (even one running on a Spectrum) would support VT100. Locally it would deliver VT52+color because that's what the console does. Local game is a bit of TBD. Couch-coop with "time warp" comes to mind. A halfway-capable computer player would be a nice option, too, of course.

**Comment by stefanhaustein** [comment]
But then how would another spectrum next connect if it only supports VT52? 

How about VT52 (for CP/M clients), VT52🌈, and ANSI?

The terminal include should be quite small and I don't really see any significant advantage of supporting CRT, too?
